### ISSUE TITLE
No persistence to localStorage even when store is updated in Redux Devtools

### ISSUE THREAD

**Issue by st-clair-clarke** [zuriscript/signalstory#29]
Hi,
I have the following typescript code that is being used in an Angular v16.x.x enviroment:

**enviroment.development.ts**
```
export const environment = {
   production: false,
   defaultStorePlugins: () => [
      useDevtools(),
      useStoreHistory(),
//      useStorePersistence({ persistenceKey: 'RegistrationStore' }),
      // useDeepFreeze(),
   ],
}
```

**registration-store.service.ts**
```
@Injectable({ providedIn: 'root' })
export class RegistrationStore extends ImmutableStore<TRegistrationState> {
   constructor() {
      super({
         name: 'RegistrationStore',
         mutationProducerFn:
            produce as unknown as MutationFn<TRegistrationState>,
         // enableLogging: !environment.production,
         initialState: new RegistrationState(),
         plugins: [
            useStorePersistence(
                { persistenceKey: 'RegistrationStore',
                  persistenceStorage: localStorage}),
            ...environment.defaultStorePlugins(),
         ],
      })
   }

   store: Immutable<TRegistrationState> = this.state()

   patientName = (): TName => computed(() => this.patient().name)()
   updatePatientName = (name: TName) => {
      this.update(
         (state) =>
            produce(state, (draft) => {
               draft.domain.patient.name = name
            }),

         createAction({ parent: 'patient', children: ['name'] }),
      )
   }

... other code

```

The updatePatientName() updates the store successfully as seen in the Redux Devtools screenshot, but the localStore fails to update? I would like your help in solving this issue.

![image](https://github.com/zuriscript/signalstory/assets/3227590/8c542d90-f312-461a-ac37-1a641ce3546f)

Thanks

**Comment by zuriscript** [comment]
Hi there

I have tried to reproduce this behaviour, but it seems to work.
Here is a minimal working example based on your code and Angular 16.2.12 and signalstory 0.1.2: [Stackblitz Sample](https://stackblitz.com/edit/stackblitz-signalstory-samples-issue-29?file=src%2Fmain.ts)

I suspect that the error resides in another place in your code.
Can you sort out any other errors? Have you tried multiple Browser? Have you tried printing the current value (and after updating the state) of localStorage to Console to make sure this is not a bowser issue?

**Comment by zuriscript** [comment]
@st-clair-clarke, reconsidering, I highly doubt this is browser-related.

I suspect it's tied to how effects function in Angular 16. According to the docs, "Effects always execute asynchronously, during the change detection process." So far, effects only run during a change detection cycle. If you happened to change a store value after change detection, the state change wouldn't have been propagated to local storage using the effects mechanism.

This could be related to [https://angular.io/errors/NG0100](ExpressionChangedAfterItHasChangedError) and might point to a bug in your code. Nevertheless, I've chosen to switch to synchronous local storage updates for now to support any use case. Therefore, I've released version 0.1.3 of signalstory.

Considering the change in behavior (https://github.com/angular/angular/pull/51049), I'll reevaluate for Angular 17 and possibly switch back to asynchronous local storage updates. A side note: I'll soon release the first stable/productive version and will continue semantic versioning from there, aligned with Angular versions.

Can you try your code with signalstory@0.1.3?

**Comment by st-clair-clarke** [comment]
Version 0.1.3 works perfect! Once it was installed everything functions the way it was originally. It was an update in the past of signalstory that broke the functionality. Looking forward to the production version - just in time with angular 17!

This is the easiest state management tool I have seen and used. Congratulations again on an excellent library.

**Comment by zuriscript** [comment]
Appreciate your kind words.

But to clarify, it is not a new bug (or necessaraly a bug at all). The persistence plugin's behavior has stayed the same since the beginning and hasn't changed. It always relied on [signal effects](https://angular.io/guide/signals#effects) to react to changes in the store state. This means it won't work if change detection isn't running, like for example inside components using ChangeDetectionStrategy.OnPush.

Rest assured, I've only used signal effects in this specific case.

In Angular 17, effects will be scheduled by the microtask queue, solving this issue. 
For Signalstory, I've decided to do this synchronously, in order to solve the problem for you but also to get more resilient. After each command, the sync to storage happens immediately, avoiding reliance on change detection. I might go back to signal effects for this and other things if it's safe, but signal effects remain a delicate topic for sure.
